Allows user code to retain references to memoized helpers

JoshCheek opened the pull request:
Accomplish this by no longer clearing the example's ivars.

Fixes rspec/rspec-core#1921

Context
-------

* Originally clearing ivars due to memory leak:
  https://github.com/rspec/rspec-core/issues/321
* Threadsafe memoized helpers caused `__memoized` to stop lazily initializing:
  https://github.com/rspec/rspec-core/issues/321
* This caused it to be permawiped by by the resetting of the example's ivars:
  https://github.com/rspec/rspec-core/issues/1921

However, this patch tests the behaviour of the memory leak,
rather than its mechanics, which shows that it was fixed at some point.
So we simply remove that code.

JoshCheek commented:
Looks like this CI issue is hitting the other PRs, too. It's strange, all the paths are in there, but they're expanded differently or in different locations.

It does look like the 1.8.7 build is actually failing, though: https://travis-ci.org/rspec/rspec-core/jobs/60327101#L397 Is there a way to omit it only for a given Ruby? I assume if it passes on 2.1+ that it can be assumed fine on the others, even if I don't know how to explicitly start their gc.

myronmarston commented:
> Looks like this CI issue is hitting the other PRs, too. It's strange, all the paths are in there, but they're expanded differently or in different locations.

It's not hitting other PRs, but we do run the spec suites (but not the cukes) from the other repos as part of our CI builds, to ensure a change in one repo doesn't break the build in another repo.  In this case, there's a spec in rspec-support that is failing for an unknown reason.  I noticed the same failure in @JonRowe's PR in rspec/rspec-support#199.  I'm not sure why yet. I'll go push a change to that repo to skip that spec for now until we can figure out why that's happening.

> It does look like the 1.8.7 build is actually failing, though: https://travis-ci.org/rspec/rspec-core/jobs/60327101#L397 Is there a way to omit it only for a given Ruby? I assume if it passes on 2.1+ that it can be assumed fine on the others, even if I don't know how to explicitly start their gc.

You can prevent the example from being executed by tagging it with `:if => (RUBY_VERSION.to_f > 1.8)`, although I think it might be worth investigating it a bit before we do that.  The failure makes it look like there are 2 of the objects that should have been GC'd that were not.  That suggests there may still be a memory leak on 1.8.7.

JoshCheek commented:
Oh, I assumed it meant that `GC.start` was merely a suggestion, and it didn't necessarily run GC. But maybe that's nonsense, and I was really thinking of, thread scheduling commands or something. I'll play with it and see if I can't get it locally (I was able to get it installed, though it's pretty rough to work with).

myronmarston commented:
I did some troubleshooting and it looks like it's the `:before_all` and `:after_all` garbage instances that aren't being GC'd.  I'm not sure why.  I also did `git checkout master -- lib` to see if the same thing happens on master, and it does.  So it appears that ivars declared in `:all` hooks might leak on 1.8.7 for some reason, and this was already happening (so it's not a result of your PR).  If you want to do more digging, feel free (it would be much appreciated!), or you can tweak the spec to work slightly differently on 1.8.7 (e.g. skip the `:all` hooks on 1.8.7).

While digging around, I also noticed that it wasn't always 8 instances in `ObjectSpace` before the `run_gc` call.  Since Ruby can run GC at literally anytime, it can run it before we measure the "before GC garbage count", causing some indeterminism in the test.  Can you disable GC during the test (apart from your manual GC) so it's deterministic?

One last suggestion: what do you think about deleting [this example](https://github.com/rspec/rspec-core/blob/9f36715d891d41faf8b7db168f9e64b59157f9cb/spec/rspec/core/example_spec.rb#L254-L260)?  IMO, that behavior isn't interesting on its own -- it's the lack of a memory leak that we care about.  Maybe we'll find a better way to prevent the memory leak in the future that doesn't involve clearing the `example_group_instance` ivar, and it'd be nice to not have that spec making us think that's an important behavior.

Thanks, @JoshCheek, this is looking really good!  I'm impressed with how quickly you put this together :).

JonRowe commented on lib/rspec/core/example.rb:
```diff
@@ -225,10 +225,7 @@ def run(example_group_instance, reporter)
-            @example_group_instance.instance_variable_set(ivar, nil)
-          end
-          @example_group_instance = nil
+          @example_group_instance = nil # if you love something... let it go
```
Heh, I think the comment is probably unnecessary ;)

myronmarston replied:
But it makes @JoshCheek super happy!

https://twitter.com/josh_cheek/status/592937805636050944

JonRowe replied:
Heh... still....

JoshCheek replied:
I'll remove it if you explicitly tell me to.

myronmarston replied:
I don't care either way.  Personally, I like when people put a little personality in their commits, so I'm happy to let it stay.

JoshCheek commented:
Do you squash each commit and then force push it? I'm not sure I'm doing the single-commit workflow correctly.

> I did some troubleshooting and it looks like it's the :before_all and :after_all garbage instances that aren't being GC'd. I'm not sure why. I also did git checkout master -- lib to see if the same thing happens on master, and it does. So it appears that ivars declared in :all hooks might leak on 1.8.7 for some reason, and this was already happening (so it's not a result of your PR). If you want to do more digging, feel free (it would be much appreciated!), or you can tweak the spec to work slightly differently on 1.8.7 (e.g. skip the :all hooks on 1.8.7).

I got it running locally on 1.8.7, and I'm pretty confident it releases all references, I think it's an issue with the GC triggering.

I fixed the flakiness from not disabling the GC, so it always has 8 beforehand, but it still usually has 2 afterwards, they're the before and after all, as you mentioned. Sometimes that number varies, though.

When I place a pry in there, it goes down to zero. Without the pry, I usually get the 2 from the context hooks.

> One last suggestion: what do you think about deleting this example? IMO, that behavior isn't interesting on its own -- it's the lack of a memory leak that we care about. Maybe we'll find a better way to prevent the memory leak in the future that doesn't involve clearing the example_group_instance ivar, and it'd be nice to not have that spec making us think that's an important behavior.

Yeah, good call, I checked that the test I added fails if this behaviour isn't in place, I'll delete it.

> Thanks, @JoshCheek, this is looking really good! I'm impressed with how quickly you put this together :).

Aww, thx ^_^ I think it was already fixed, just the test was still here, so all I have to do is come up w/ a test that reproduces the issue, then show that it passes.

myronmarston commented:
> Do you squash each commit and then force push it? I'm not sure I'm doing the single-commit workflow correctly.

Depends...if the change logically belongs with an existing commit, I'll squash it in with that commit.  If I think it belongs as its own commit (and particularly if I think that future history spelunking could benefit from the change being in its own commit with an extended explanation in the commit message) I'll leave it as its own commit.

My general litmus test is, "when I'm digging into the history of these changes in the future to understand why it got into this state, will be the most useful commit organization" -- and I aim to organize my commits that way within reason.  Sometimes it would be a lot of work to do that so I don't.

> I got it running locally on 1.8.7, and I'm pretty confident it releases all references, I think it's an issue with the GC triggering.

> I fixed the flakiness from not disabling the GC, so it always has 8 beforehand, but it still usually has 2 afterwards, they're the before and after all, as you mentioned. Sometimes that number varies, though.

> When I place a pry in there, it goes down to zero. Without the pry, I usually get the 2 from the context hooks.

Weird.  I think it's reasonable to either skip that test on 1.8.7 or use some version-based conditionals to get it to pass with whatever behavior is consistent.

myronmarston commented:
Thanks, @JoshCheek!

myronmarston commented:
Sadly, we just got a failure from the "does not leak memory" spec added in this PR:

https://travis-ci.org/rspec/rspec-core/jobs/61159415

```
  1) RSpec::Core::Example#run memory leaks, see GH-321, GH-1921 releases references to the examples / their ivars
     Failure/Error: expect {
       expected result to have changed to 0, but is now 1
```

I have no idea why it is failing.  The GC seems like it should be deterministic with the fact that it is disabled and then re-enabled and manually started with specific options.

@JoshCheek -- any ideas what could be causing that?  The job above failed on `ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]` so it's a pretty recent ruby version.

myronmarston commented:
Hmm....one thought: is `GC.start` synchronous?  If it was async it could make sense that it could occasionally fail.

JoshCheek commented:
Prob won't get back to this until the weekend or next week, so figured I should post what I have at the moment:

It's alarming that it passed on this same version, on the test that pulled it in: https://travis-ci.org/rspec/rspec-core/jobs/60615388 Intermittent failures are the worst :(

When I was doing research, I found it pretty difficult to force GC to actually happen. This is why it uses the `reliable_gc` conditional https://github.com/rspec/rspec-core/blob/c712c90c5855ac0a675830391a5696525bda829a/spec/rspec/core/example_spec.rb#L347

In the end, I decided that `GC.start` was simply unreliable without being able to tell it to perform the sweep right in that moment. [Here](http://ruby-doc.org/core-2.1.5/GC.html#method-c-start) are the docs for it on 2.1.5. So, unless I misunderstand it (which is not unreasonable), passing these options should cause it to perform the full GC right in that moment.

Here is an experiment I ran, and GC does not appear to be threadsafe, but I'm not sure of a good way to tell if its running asynchronously. But the non-threadsafe, alone, is enough to mess with GC (eg enabling would allow it to clean up some of the objects, disabling would prevent it from cleaning up the last one)

```ruby
# From http://www.rubydoc.info/stdlib/core/GC
# GC.enable  -- Enables garbage collection, returning true if garbage collection was previously disabled.
# GC.disable -- Disables garbage collection, returning true if garbage collection was already disabled.

RUBY_VERSION # => "2.2.2"

t = Thread.new do
  GC.disable # => false
  sleep 1
  GC.enable  # => true
end

sleep 0.5
GC.disable # => true
t.join
GC.enable  # => false
```

My next plan was to look through all the installed gems for accessing of GC and ObjectSpace, to see if they could be messing with it. Perhaps monkey patch the GC methods to record all the callsites. I had started doing this on Monday (just grepping in ./bundle/ruby/2.1.0) and it looks like there's a lot of comments in Nokogiri about how to get GC to behave like we're doing. So I'll probably look at how they do it, see if I can't learn something from it.

If it turns out there's something else messing with GC, there are a few options, everything from addressing that situation specifically, to swapping the GC constant with a threadsafe proxy. Not super stoked about that idea, but at least we wouldn't wind up in this spot again, with some other dep added later.

Of course, it's also possible that I just don't understand GC sufficiently.

myronmarston commented:
Thanks for the follow up @JoshCheek.  I really appreciate your willingness to dig into this issue!

If you reach a point where you can't figure figure out why it's non-deterministic and don't have more time to devote to it, one potential solution would be something like this:

``` ruby
def force_gc_until
  GC.enable
  10.times do
    GC.start :full_mark => true, :immediate_sweep => true
    return if yield
    sleep 0.001
  end
end

# in the example it would do:
expect(ObjectSpace.each_object(garbage).map(&:defined_in).map(&:to_s).sort).to eq(%w[
  after_all  after_each  after_each
  before_all before_each before_each
  failing_example passing_example
])

force_gc_until { ObjectSpace.each_object(garbage).count.zero? }

expect(ObjectSpace.each_object(garbage).map(&:defined_in).map(&:to_s).sort).to eq([])
```

Basically, run the GC repeatedly up to a set number of times.  While not ideal, if this test consistently passes than the behavior we care about is satisfied: it demonstrates that the GC will eventually collect these objects even if it didn't necessarily happen on the first GC.  Whether or not it consistently happens on the first GC is really an implementation detail of the GC and not something we should care to test.

Actually, I have some vague memories of reading something about how Ruby's GC may not collect all collectable objects on the first GC after they are collectable, because it limits its object scan to a single pass for any given GC, so on the object scan of the first GC it may simply put it on a "possibilities" list and than on the 2nd GC it would actually collect it.  (As you can see, my understanding of Ruby's GC is pretty limited as well...)  Anyhow, it may be that we just have to trigger 2 GC runs to make this deterministic.  Given how cheap and easy it would be to do that, we could make that change and just see if the test ever fails again.  Thoughts?

JoshCheek commented:
The Installed Gems
--------------------------

Looked through the installed gems, none of them look to be messing with this. Several were dealing with similar issues, though. Here's what I see:

* FFI has a method called [force_gc](https://github.com/ffi/ffi/blob/d9db892b99b55449621acf1525a5e6ede06b7aff/spec/ffi/spec_helper.rb#L82-89) which looks at a bunch of different things to determine how to run it. Might be useful, esp if it starts failing on Rbx or JRuby. Note that Nokogiri has a very [similar method](https://github.com/sparklemotion/nokogiri/blob/34b2a245578291e251b2d07ed57cb065c443659d/test/helper.rb#L39-46).
* Nokogiri does basically what we do, and they have explicitly [lowered the strictness of the test](https://github.com/sparklemotion/nokogiri/commit/35544d9309e221c69d2a78e545f6c5ae6bcd444e) to give themselves "wiggle room". Also note that they run GC twice in that test.
* FFI also gives themselves [wiggle room](https://github.com/ffi/ffi/blob/d9db892b99b55449621acf1525a5e6ede06b7aff/spec/ffi/pointer_spec.rb#L145), with a comment saying "GC rarely cleans up all objects. we can get most of them, and that's enough to determine if the basic functionality is working."
* FFI has a method [wait_gc](https://github.com/ffi/ffi/blob/d9db892b99b55449621acf1525a5e6ede06b7aff/spec/ffi/fixtures/classes.rb#L138-149) which loops until GC has collected what they expect. This is similar to the `force_gc_until` method @myronmarston suggests above.

-----

Recording all GC/ObjectSpace calls
------------------------------------------------

So far, I don't see anything that would obviously cause this. I recorded all method calls on GC and ObjectSpace with this code:

```ruby
class ISeeWhatYouDidThere < BasicObject
  def initialize(name, target, stream)
    @name, @target, @stream = name, target, stream
  end

  def method_missing(methodname, *args, &block)
    result = @target.__send__(methodname, *args, &block)
    @stream.puts "From #{invoker}"
    @stream.puts "  #{@name}.#{methodname}(#{args.map(&:inspect).join(', ')})#{' { ... }' if block}"
    @stream.puts "  # => #{result.inspect}"
    @stream.puts
    result
  end

  private

  def invoker
    ::Object.new.__send__(:caller)[1]
  end
end

filename = File.expand_path '../../debug.log',  __FILE__
logfile  = File.open filename, 'a'
at_exit { logfile.close }
GC          = ISeeWhatYouDidThere.new 'GC',          GC,          logfile
ObjectSpace = ISeeWhatYouDidThere.new 'ObjectSpace', ObjectSpace, logfile
```

The calls are listed below. They don't look like they would mess with this.

So, I'm inclined to suspect that it's just difficulty getting the GC to be reliable (it would be really nice if ObjectSpace included some way to assert that an object was unreachable -- [best I've seen](http://www.rubydoc.info/stdlib/objspace/ObjectSpace#reachable_objects_from-class_method) would require a lot of work and bringing in a stdlib dependency).

Since you thought of the loop, and so did FFI, I'm inclined to go with that. It's straightforward, and probably makes the issue negligibly small. I definitely like it better than monkey patching GC.

```
From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:344:in `reliable_gc'
  GC.method(:start)
  # => #<Method: GC.start>

From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:344:in `reliable_gc'
  GC.method(:start)
  # => #<Method: GC.start>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824fc59288 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824fc59288 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824fa831e8 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824fa831e8 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824fa10c38 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824fa10c38 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824f969960 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824f969960 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824f902828 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824f902828 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824f8599a8 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824f8599a8 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824f80ab50 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824f80ab50 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f8250686800 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f8250686800 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:240:in `unlink'
  ObjectSpace.undefine_finalizer(#<Tempfile:>)
  # => #<Tempfile:>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f8250697650 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f8250697650 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:240:in `unlink'
  ObjectSpace.undefine_finalizer(#<Tempfile:>)
  # => #<Tempfile:>

From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:366:in `block (4 levels) in <top (required)>'
  GC.disable()
  # => false

From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:373:in `block (5 levels) in <top (required)>'
  ObjectSpace.each_object(#<Class:0x007f8251a70028>)
  # => #<Enumerator: ObjectSpace:each_object(#<Class:0x007f8251a70028>)>

From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:370:in `block (5 levels) in <top (required)>'
  GC.enable()
  # => true

From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:371:in `block (5 levels) in <top (required)>'
  GC.start({:full_mark=>true, :immediate_sweep=>true})
  # => nil

From /Users/josh/ref/ruby/rspec-core/spec/rspec/core/example_spec.rb:373:in `block (5 levels) in <top (required)>'
  ObjectSpace.each_object(#<Class:0x007f8251a70028>)
  # => #<Enumerator: ObjectSpace:each_object(#<Class:0x007f8251a70028>)>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824da8b340 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824da8b340 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098847800640)
  # => #<StringIO:0x007f8251d83a80>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098847800600)
  # => #<StringIO:0x007f8251d83a30>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098848227160)
  # => #<StringIO:0x007f8251e53eb0>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098848227120)
  # => #<StringIO:0x007f8251e53e60>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098848802400)
  # => #<StringIO:0x007f8251f6ccc0>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098848398780)
  # => #<StringIO:0x007f8251ea7b78>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098849617520)
  # => #<StringIO:0x007f82520face0>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/drb/drb.rb:369:in `to_obj'
  ObjectSpace._id2ref(70098849617480)
  # => #<StringIO:0x007f82520fac90>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824c418838 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824c418838 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f824c7bbf80 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f824c7bbf80 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:240:in `unlink'
  ObjectSpace.undefine_finalizer(#<Tempfile:>)
  # => #<Tempfile:>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:240:in `unlink'
  ObjectSpace.undefine_finalizer(#<Tempfile:>)
  # => #<Tempfile:>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f8251968fb8 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f8251968fb8 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:134:in `initialize'
  ObjectSpace.define_finalizer(#<Tempfile:>, #<Tempfile::Remover:0x007f825198b1a8 @pid=78571, @data=[]>)
  # => [0, #<Tempfile::Remover:0x007f825198b1a8 @pid=78571, @data=[]>]

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:240:in `unlink'
  ObjectSpace.undefine_finalizer(#<Tempfile:>)
  # => #<Tempfile:>

From /Users/josh/.rubies/ruby-2.1.1/lib/ruby/2.1.0/tempfile.rb:240:in `unlink'
  ObjectSpace.undefine_finalizer(#<Tempfile:>)
  # => #<Tempfile:>
```

myronmarston commented:
Thanks for following upon this, @JoshCheek!

> Since you thought of the loop, and so did FFI, I'm inclined to go with that. It's straightforward, and probably makes the issue negligibly small. I definitely like it better than monkey patching GC.

:+1:

It's too bad the GC's behavior isn't deterministic :(.
